File: lua/armstrong-numbers/armstrong-numbers.lua
local ArmstrongNumbers <const> = {}

local armstrong_sum, armstrong_digits_transformer;
local digits_of, vec_sum, vec_map;

--- Checks if a given integer is an Armstrong (narcissistic) number.
--- @param number integer The number to evaluate.
--- @return boolean # True if the number equals the sum of its digits raised to the power of the total digit count.
function ArmstrongNumbers.is_armstrong_number(number)
    local arm_sum <const> = armstrong_sum(number);

    return arm_sum == number;
end

--- Creates a transformation function that raises a digit to the power of the total digit count.
--- @param num_size integer The total number of digits.
--- @return fun(digit: integer): integer # A function that raises an input digit to the specified power.
function armstrong_digits_transformer(num_size)
    return function(d)
        return d ^ num_size;
    end
end

--- Calculates the Armstrong sum for a given integer.
--- @param num integer The integer to process.
--- @return integer # The sum of each digit raised to the power of the digit count.
function armstrong_sum(num)
    local digits <const> = digits_of(num);
    local arm_power = armstrong_digits_transformer(#digits);
    local arm_digits <const> = vec_map(digits, arm_power);

    return vec_sum(arm_digits);
end

--- Extracts the individual base-10 digits of an integer in order.
--- @param num integer The integer to decompose.
--- @return integer[] # An array containing the digits of the number.
function digits_of(num)
    local current_num = num;
    local digits <const> = {};

    while current_num > 0 do
        local digit = current_num % 10;

        table.insert(digits, 1, digit);

        current_num = math.floor(current_num / 10);
    end

    return digits;
end

--- Calculates the sum of all elements in a numeric array.
--- @param vec number[] The array of numbers to sum.
--- @return number # The total sum of the elements.
function vec_sum(vec)
    local sum = 0;
    for _, v in ipairs(vec) do
        sum = sum + v;
    end

    return sum;
end

--- Maps each element of an array using a transformation function.
--- @generic T, U
--- @param vec T[] The input array to transform.
--- @param op fun(elem: T): U The transformation function.
--- @return U[] # A new array containing the mapped values.
function vec_map(vec, op)
    local mapped <const> = {};

    for i, v in ipairs(vec) do
        mapped[i] = op(v);
    end

    return mapped;
end

return ArmstrongNumbers;

File: php/armstrong-numbers/ArmstrongNumbers.php
<?php

declare(strict_types=1);

/**
 * Checks if a given number is an Armstrong number.
 *
 * An Armstrong number (also known as a narcissistic number, pluperfect digital invariant,
 * or Plus Perfect Number) is a number that is the sum of its own digits each raised
 * to the power of the number of digits.
 *
 * @param int $number The integer to check.
 * @return bool True if the number is an Armstrong number, false otherwise.
 *
 * @example
 * isArmstrongNumber(9);   // Returns true (9^1 = 9)
 * isArmstrongNumber(153); // Returns true (1^3 + 5^3 + 3^3 = 1 + 125 + 27 = 153)
 * isArmstrongNumber(10);  // Returns false (1^2 + 0^2 = 1 != 10)
 */
function isArmstrongNumber(int $number): bool
{
    return armstrongSum($number) === $number;
}

/**
 * Calculates the Armstrong sum for a given number.
 *
 * The Armstrong sum is the sum of each digit of the number raised to the power
 * of the total number of digits in the original number.
 *
 * @param int $number The integer for which to calculate the Armstrong sum.
 * @return int The calculated Armstrong sum.
 */
function armstrongSum(int $number): int
{
    $digits = digits($number);
    $digitsCount = count($digits);
    $sum = array_sum(array_map(fn(int $digit) => $digit ** $digitsCount, $digits));

    return (int) $sum;
}

/**
 * Extracts individual digits from an integer and returns them as an array.
 *
 * The digits are returned in their original order (from most significant to least significant).
 *
 * @param int $number The integer from which to extract digits.
 * @return array<int> An array of integers, where each element is a digit of the original number.
 *
 * @example
 * digits(153); // Returns [1, 5, 3]
 * digits(9);   // Returns [9]
 * digits(0);   // Returns []
 */
function digits(int $number): array
{
    $digits = [];

    for ($current = $number; $current > 0; $current = (int) ($current / 10)) {
        array_unshift($digits, $current % 10);
    }

    return $digits;
}